src/mast/mod.rs

```diff
@@ -0,0 +1,1108 @@
+                to_delete.push(name.clone());
+            }
+        }
+        for d in to_delete {
+            self.vars.remove(&d);
+        }
+    }
+
+    /// Returns true if a scope is a prefix of our scope.
+    pub fn is_in_scope(&self, prefix_scope: usize) -> bool {
+        self.current_scope >= prefix_scope
+    }
+
+    /// Stores type information about a local variable.
+    pub fn store_type(&mut self, ident: &str, type_info: &MTypeInfo) -> Result<()> {
+        match self
+            .vars
+            .insert(ident.to_string(), (self.current_scope, type_info.clone()))
+        {
+            Some(_) => Err(Error::new(
+                "type-checker",
+                ErrorKind::DuplicateDefinition(ident.to_string()),
+                type_info.span,
+            )),
+            None => Ok(()),
+        }
+    }
+
+    /// Retrieves type information on a variable, given a name.
+    /// If the variable is not in scope, return None.
+    pub fn get_type_info(&self, ident: &str) -> Option<&MTypeInfo> {
+        if let Some((scope, type_info)) = self.vars.get(ident) {
+            if self.is_in_scope(*scope) {
+                Some(type_info)
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+}
+
+/// A context to store the last node id for the monomorphized AST.
+#[derive(Debug, Default)]
+pub struct MastCtx {
+    pub last_node_id: usize,
+}
+
+impl MastCtx {
+    pub fn new() -> Self {
+        Self { last_node_id: 0 }
+    }
+
+    pub fn next_node_id(&mut self) -> usize {
+        self.last_node_id += 1;
+        self.last_node_id
+    }
+}
+
+#[derive(Debug)]
+/// Mast relies on the TAST for the information about the "unresolved" types to monomorphize.
+/// Things such as loading the function AST and struct AST from fully qualified names.
+/// After monomorphization process, the following data will be updated:
+/// - Resolved types. This can be used to determine the size of a type.
+/// - Instantiated functions. The circuit writer will load the instantiated function AST by node id.
+/// - Monomorphized AST is generated for the circuit writer to walk through and compute.
+pub struct Mast<B>
+where
+    B: Backend,
+{
+    tast: TypeChecker<B>,
+
+    /// Mapping from node id to resolved type
+    node_types: HashMap<usize, TyKind>,
+
+    /// Mapping from node id to instantiated functions
+    node_functions: HashMap<usize, FnInfo<B>>,
+
+    ctx: MastCtx,
+
+    /// Monomorphized AST of the main function
+    pub main_fn_ast: Option<FunctionDef>,
+}
+
+impl<B: Backend> Mast<B> {
+    pub fn new(tast: TypeChecker<B>) -> Self {
+        Self {
+            tast,
+            node_types: HashMap::new(),
+            node_functions: HashMap::new(),
+            ctx: MastCtx::new(),
+            main_fn_ast: None,
+        }
+    }
+
+    pub fn expr_type(&self, expr: &Expr) -> Option<&TyKind> {
+        self.node_types.get(&expr.node_id)
```

Comment:
right, the `TAST::node_types` should be removed. `MAST::node_types` is regenerated during the monomorphization, as node ids are regenerated.

Reply:
I think we can just keep TAST::node_types as it is, since circuit writer just relies on this regenerated node_types from MAST

Reply:
so I'm still not sure I understand why we are recreating node IDs for every nodes. My current undersatnding is that:

1. when we monomorphize generic functions, we create different versions for each instantiation, and thus we have different functions with the same nodes in their AST
2. If we want to include information in these nodes that's **specific to the instantiated generic function** then we need to give them a new ID
3. If we didn't assign specific information to these nodes, then it would be fine if they shared the same ID
4. currently, we don't assign specific information, but we chose to in the monomorphization phase
5. if we modified the AST with the specific information (which we are already doing anyway), we would not need to assign new node IDs

is this correct?

Reply:
added a new section in the rfc to explain why new node ids are needed